crates/ra_hir/src/path.rs

```diff
@@ -76,14 +76,32 @@ fn expand_use_tree(
 ) {
     if let Some(use_tree_list) = tree.use_tree_list() {
         let prefix = match tree.path() {
+            // E.g. use something::{{{inner}}};
             None => prefix,
+            // E.g. `use something::{inner}` (prefix is `None`, path is `something`)
+            // or `use something::{path::{inner::{innerer}}}` (prefix is `something::path`, path is `inner`)
             Some(path) => match convert_path(prefix, path) {
                 Some(it) => Some(it),
                 None => return, // TODO: report errors somewhere
             },
         };
-        for tree in use_tree_list.use_trees() {
-            expand_use_tree(prefix.clone(), tree, cb);
+        for child_tree in use_tree_list.use_trees() {
+            // Handle self in a path.
+            // E.g. `use something::{self, <...>}`
+            if let Some(path) = child_tree.path() {
+                if path.qualifier().is_none() {
+                    if let Some(segment) = path.segment() {
+                        if segment.kind() == Some(ast::PathSegmentKind::SelfKw) {
+                            /* TODO: Work out what on earth range means in this callback */
+                            if let Some(prefix) = prefix.clone() {
+                                cb(prefix, Some(segment.syntax().range()));
+                                continue;
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
Yep, I think this'll work. Can we, howver, move this logic to this line?

https://github.com/rust-analyzer/rust-analyzer/blob/d8d60215dacc3eeddeda4d6af7007e2129fdb00e/crates/ra_hir/src/path.rs#L109

that way, you'll need to check that path.kind is Self and that segments are empty. This should be easier then these nested ifs I suppose?

Reply:
Hm, can't we reuse `convert_path`? what we are doing here is determining if a fragment starts with `self,`, but `conver_path` also has this logic I think?

Reply:
So this is one of the cases where `convert_path` returns `None`. I would still have to check before [this translation](https://github.com/DJMcNab/rust-analyzer/blob/bb1ee2f13a67342bdb09bcbf8684f6c9627b89e1/crates/ra_hir/src/path.rs#L121) to avoid duplicating the work (creating a new `prefix` then throwing it away again), which would have the same body as the current implementation.